### DISCUSSION TITLE
Resolver Function to JSON Filename

### DISCUSSION THREAD

**Discussion by 0xc0de4c0ffee** [namesys-eth/ccip2-eth-resources#5] [Ideas]
Resolver function to json file name
| | Type | Function | Json file | 
| -- | -- | -- | --- |
|a)| Text Records (ENSIP-05) | `text(bytes32 node, string memory key)` | ../\<key>.json |
|b)| Ethereum address | `../addr(bytes32 node)` | addr-60.json | 
|c)|Multichain Address (ENSIP-9) | `addr(bytes32 node, uint coinType)`| ../addr-\<coinType>.json |
|d)|Public Key | `pubkey(bytes32 node)`| ../pubkey.json | 
|e)|*Contenthash (ENSIP-07) | `contenthash(bytes32 node)` | ../contenthash.json |
|f)|Interface (ENSIP-08) | `interfaceImplementer(bytes32 node, bytes4 interfaceID)` | ../interface/0x\<interfaceID>.json |
|g)|Zonehash (ENSIP-06) | `zonehash(bytes32 node)` | ../dns/zonehash.json | 
|h)|DNS Records (ENSIP-06) | `dnsRecord(bytes32 node, bytes32 name, uint16 resource)` | ../dns/0x\<name>-\<resource>.json | 
|i)|ABI (ENSIP-04) | `ABI(bytes32 node, uint256 contentTypes)` | abi-\<contentTypes>.json | 

* contenthash isn't used as off-chain record as it's required to be on-chain to resolve other records.

**Comment by 0xc0de4c0ffee** [comment]
wip: https://github.com/namesys-eth/ccip2-eth-resolver/pull/3 

json path `gateway.tld/ipns/f<hash>/.well-known/eth/domain/...`
update (b)&(c) to `../_addr/<cointype>.json`
all dirs are now `_dir` format to stop collision with "dir" as subdomain.

```
    if (bytes(funcToFile[func]).length > 0) {
        _pathJSON = funcToFile[func];
    } else if (func == iResolver.text.selector) {
        _pathJSON = abi.decode(data[36:], (string));
    } else if (func == iResolver.addr.selector) {
        _pathJSON = string("_addr/60");
    } else if (func == iOverloadResolver.addr.selector) {
        _pathJSON = string.concat("_addr/", uintToNumString(abi.decode(data[36:], (uint256))));
    } else if (func == iResolver.interfaceImplementer.selector) {
        _pathJSON =
            string.concat("_interface/", bytes2HexString(abi.encodePacked(abi.decode(data[36:], (bytes4))), 0));
    } else if (func == iResolver.ABI.selector) {
        // recheck this
        _pathJSON = string.concat("_abi/", uintToNumString(abi.decode(data[36:], (uint256))));
    } else if (func == iResolver.dnsRecord.selector) {
        (bytes32 _name, uint16 resource) = abi.decode(data[36:], (bytes32, uint16));
        _pathJSON =
            string.concat("_dns/", bytes2HexString(abi.encodePacked(_name), 0), "/", uintToNumString(resource));
    } else {
        revert NotImplemented(func);
    }
```

**Reply by 0xc0de4c0ffee** [reply]
`bytes2HexString` in ../path/.. should be `0x` prefixed? or leave it as \<hex> only??
*except for _ipns contenthash, it's "f"prefixed hex string.

**Reply by sshmatrix** [reply]
> `bytes2HexString` in ../path/.. should be `0x` prefixed? or leave it as <hex> only?? *except for _ipns contenthash, it's "f"prefixed hex string.

Don't really know if it makes a difference... 😶‍🌫️  Is there a standard going around? ENS likes to do `0x` a lot

**Reply by 0xc0de4c0ffee** [reply]
ok, I'll add `0x` prefix.. 
single "_" or double "__" for dir prefix? 
we're using "__callback" too.. "_single" is good enough.
_addr or _address? 

I was also checking SLIP44 to see if we could use "../_address/eth.json" instead of "../_address/60.json".. but slip44 list is too long = not $cheap to add/manage all on chain records.

**Reply by sshmatrix** [reply]
Choose as you like 🖖
